tests/test_inference_problem.py

```diff
@@ -0,0 +1,437 @@
+"""
+
+
+class MySensor(Sensor):
+    def __init__(self, name, position):
+        super().__init__(name, shape=(1, 1))
+        self.position = position
+
+
+class MyForwardModel:
+    def __call__(self, parameter_list, sensors, time_steps):
+        """
+        evaluates 
+            fw(x, t) = A * x + B * t
+        """
+        A = parameter_list["A"]
+        B = parameter_list["B"]
+
+        result = {}
+        for sensor in sensors:
+            result[sensor] = A * sensor.position + B * time_steps
+        return result
+
+    def parameter_list(self):
+        p = ParameterList()
+        p.define("A", None)
+        p.define("B", None)
+        return p
+
+
+class MyModelError(ModelError):
+    def __init__(self, fw, data):
+        self._fw = fw
+        self._ts, self._sensor_data = data
+
+    def __call__(self, prm):
+        sensors = list(self._sensor_data.keys())
+        model_response = self._fw(prm, sensors, self._ts)
+        error = {}
+        for sensor in sensors:
+            error[sensor] = model_response[sensor] - self._sensor_data[sensor]
+        return error
+
+
+"""
+###############################################################################
+
+                            MAIN CODE
+
+###############################################################################
+"""
+
+if __name__ == "__main__":
+    # Define the sensor
+    s1, s2, s3 = MySensor("S1", 0.2), MySensor("S2", 0.5), MySensor("S3", 42.0)
+
+    fw = MyForwardModel()
+    prm = fw.parameter_list()
+
+    # set the correct values
+    A_correct = 42.0
+    B_correct = 6174.0
+    prm["A"] = A_correct
+    prm["B"] = B_correct
+
+    np.random.seed = 6174
+    noise_sd1 = 0.2
+    noise_sd2 = 0.4
+
+    def generate_data(N_time_steps, noise_sd):
+        time_steps = np.linspace(0, 1, N_time_steps)
+        model_response = fw(prm, [s1, s2, s3], time_steps)
+        sensor_data = {}
+        for sensor, perfect_data in model_response.items():
+            sensor_data[sensor] = perfect_data + np.random.normal(
+                0.0, noise_sd, N_time_steps
+            )
+        return time_steps, sensor_data
+
+    data1 = generate_data(101, noise_sd1)
+    data2 = generate_data(51, noise_sd2)
+
+    me1 = MyModelError(fw, data1)
+    me2 = MyModelError(fw, data2)
+
+    problem = InferenceProblem()
+    key1 = problem.add_model_error(me1, fw.parameter_list())
+    key2 = problem.add_model_error(me2, fw.parameter_list())
+
+    problem.latent.add("A", "A", key1)
+    problem.latent.add("A", "A", key2)
+    problem.latent.add_by_name("B")
+
+    problem.latent.set_prior("A", NormalDistribution(40, 5))
+    problem.latent.set_prior("B", NormalDistribution(6000, 300))
+
+    tmp = problem([A_correct, B_correct])
+
+    noise_model1 = UncorrelatedNoise()
+    noise_model2 = UncorrelatedNoise()
+    noise_key1 = problem.sensor_groups.define("exp1_noise", noise_model1)
+    noise_key2 = problem.sensor_groups.define("exp2_noise", noise_model2)
```

Comment:
Note to self.
A cleaner solution would be to almost 100% follow the way the model errors are treated. So it would read 
~~~
noise_key1 = problem.add_noise_model(noise_model1, noise_model1.define_parameter_list())
~~~
which automatically creates a sensor group (now identified by its key(!!!) instead of some name) and puts the noise parameter list to the (potentially) latent parameters.

The whole key generation should be revisited as it is essential for VB (and also helpful for the user) to distinguish between `model_error_parameters` and `noise_model_parameters`.